### PR TITLE
add user popout in navbar, add redirect-after-login, add donation goal progressbar, style tier names in account page, fix trello and discord bugs

### PR DESCRIPTION
**Pull request by ashmonty**

### CODE REVIEW

File: src/routers/account.js [hunk 1]
```diff
@@ -249,7 +254,7 @@ router.post('/login', async (request, response) => {
 
 	response.cookie('ph', encryptedBody.toString('hex'), { domain: '.pretendo.network' });
 
-	response.redirect('/account');
+	response.redirect(redirect || '/account');
```
Comment (mrjvs) [inline src/routers/account.js:257]: this seems unsafe, validate the redirect url please. (startswith / is probably sufficient)

File: src/routers/account.js [hunk 2]
```diff
@@ -268,11 +273,14 @@ router.get('/register', async (request, response) => {
 	response.clearCookie('username', { domain: '.pretendo.network' });
 	response.clearCookie('mii_name', { domain: '.pretendo.network' });
 
+	const redirect = request.query.redirect;
```
Comment (mrjvs) [inline src/routers/account.js:276]: same here, unsafe redirect

File: src/routers/account.js [hunk 3]
```diff
@@ -298,7 +306,16 @@ router.post('/register', async (request, response) => {
 	response.clearCookie('username', { domain: '.pretendo.network' });
 	response.clearCookie('mii_name', { domain: '.pretendo.network' });
 
-	response.redirect('/account');
+	response.redirect(redirect || '/account');
```
Comment (mrjvs) [inline src/routers/account.js:309]: same here, unsafe redirect

File: src/routers/blog.js [hunk 4]
```diff
@@ -69,10 +72,19 @@ router.get('/feed.xml', async (request, response) => {
...
+		postList,
+	};
 
-	const localeString = reqLocale.toString();
+	renderData.isLoggedIn = request.cookies.access_token && request.cookies.refresh_token && request.cookies.ph;
+
+	if (renderData.isLoggedIn) {
+		const account = await util.getAccount(request, response);
+		renderData.account = account;
+	}
```
Comment (mrjvs) [inline src/routers/blog.js:87]: this can be made a middleware tbh, its repeated like 50 times

**Review by mrjvs — CHANGES_REQUESTED** [review]
*(so when is this frontend gonna be written in a frontend framework) /s*

File: src/routers/blog.js (hunk 4 above)
Comment (jonbarrow) [reply src/routers/blog.js:87]: its going to be made into middleware. After all these features are merged I'm doing a big rewrite of all this auth handling

**Comment by jonbarrow** [comment]
> ((only like half kidding, the amount of logic on the frontend is growing and the backend's jank rendering code is as well))

At this point I'm not opposed to it I just don't know any frontend frameworks to be quite honest. I've never used any. But I'm curious what you mean by "backend's jank rendering"? What jank rendering? The backend just passes some basic data to the view?

**Comment by jonbarrow** [comment]
@ashmonty Forgot about this, but the home page needs to have the Patreon button changed/removed?

**Comment by ashmonty** [comment]
> @ashmonty Forgot about this, but the home page needs to have the Patreon button changed/removed?

Oh, right, good point!

How does this look?
image

**Comment by ashmonty** [comment]
(I'm adding the locale here simply because this is the only branch with the new language picker design)

**Review by mrjvs — APPROVED** [review]

**Comment by ashmonty** [comment]
> How does this look? image

@mrjvs how does this look to you, haven't committed yet

**Comment by mrjvs** [comment]
> At this point I'm not opposed to it I just don't know any frontend frameworks to be quite honest. I've never used any. But I'm curious what you mean by "backend's jank rendering"? What jank rendering? The backend just passes some basic data to the view?

its a combination of things. the mixing of data mutation and render data; the very tight data contracts between render data and stored data models; the usage of cookies for cross-page communication.

if you design a good api, the webserver would be only API. it would only be concerned between having a good data contract and that it mutates data correctly. it has much less to worry about.

a problem tho, frontend frameworks can be jank-city if they are not used correctly. having a good foundation would help tho.
jip has significant experience with frontend frameworks, you can ask them.

**Comment by mrjvs** [comment]
> @mrjvs how does this look to you, haven't committed yet

lgtm

**Review by mrjvs — APPROVED** [review]

**Comment by ashmonty** [comment]
@jonbarrow feel free to merge when ready

**Comment by jonbarrow** [comment]
> > At this point I'm not opposed to it I just don't know any frontend frameworks to be quite honest. I've never used any. But I'm curious what you mean by "backend's jank rendering"? What jank rendering? The backend just passes some basic data to the view?
> 
> its a combination of things. the mixing of data mutation and render data; the very tight data contracts between render data and stored data models; the usage of cookies for cross-page communication.
> 
> if you design a good api, the webserver would be only API. it would only be concerned between having a good data contract and that it mutates data correctly. it has much less to worry about.
> 
> a problem tho, frontend frameworks can be jank-city if they are not used correctly. having a good foundation would help tho. jip has significant experience with frontend frameworks, you can ask them.

I had originally designed the website to use an API being run on the account server (which you can still see being used for getting account details). I decided to move away from it for now because I didn’t trust a few design choices I made, and I was concerned about latency (it takes more time to execute tasks when those tasks rely on hitting a server). It’s something I’m going to revisit in the future but for now I’m going a different direction (which will be made more clear in that big cleanup PR I’ve been talking about)

**Comment by jonbarrow** [comment]
> @jonbarrow feel free to merge when ready

Sounds good!

**Pull request merged**